### PR TITLE
Collections

### PR DESCRIPTION
**Pull request by fmrico**
Hi all,

I have just finished the first step to have full support for collections.

This PR include:
* Support for collection of basic types
* Support for collection of other messages (arrays of nested)
* Tests for the above types

This PR **DOES NOT** include:
* Support for arrays of strings (you can  have an array of std_msgs/msg/String as a workaround)

Enjoy!!!

### CODE REVIEW

File: rosidl_generator_dotnet/resource/msg.cs.em [hunk 1]
```diff
@@ -115,7 +138,31 @@ public class @(type_name) : IMessage {
 
 @[for member in message.structure.members]@
 @[    if isinstance(member.type, Array)]@
-// TODO: Array types are not supported
+    private static NativeGetField@(get_field_name(type_name, member.name))MessageType native_get_field_@(member.name)_message = null;
```
Comment (theseankelly) [inline rosidl_generator_dotnet/resource/msg.cs.em:141]: For consistency?
```suggestion
    private static NativeGetField@(get_field_name(type_name, member.name))Type native_get_field_@(member.name)_message = null;
```

File: rosidl_generator_dotnet/resource/msg.cs.em [hunk 2]
```diff
@@ -115,7 +138,31 @@ public class @(type_name) : IMessage {
...
+    private static NativeGetSizeArrayField@(get_field_name(type_name, member.name))MessageType native_getsize_array_field_@(member.name)_message = null;
```
Comment (theseankelly) [inline rosidl_generator_dotnet/resource/msg.cs.em:142]: Same
```suggestion
    private static NativeGetSizeArrayField@(get_field_name(type_name, member.name))Type native_getsize_array_field_@(member.name)_message = null;
```

File: rosidl_generator_dotnet/resource/msg.cs.em [hunk 3]
```diff
@@ -159,7 +206,27 @@ public class @(type_name) : IMessage {
     public void _READ_HANDLE(IntPtr messageHandle) {
 @[for member in message.structure.members]@
 @[    if isinstance(member.type, Array)]@
-// TODO: Array types are not supported
+
```
Comment (theseankelly) [inline rosidl_generator_dotnet/resource/msg.cs.em:209]:
```suggestion
```

File: rosidl_generator_dotnet/resource/msg.cs.em [hunk 4]
```diff
@@ -159,7 +206,27 @@ public class @(type_name) : IMessage {
...
+          @[        elif isinstance(member.type.value_type, AbstractWString)]
+              // TODO: Unicode types are not supported  
+          @[        else]
+              @(get_field_name(type_name, member.name)).Add(new @(get_dotnet_type(member.type.value_type))());
+              @(get_field_name(type_name, member.name))[@(get_field_name(type_name, member.name)).Count-1]._READ_HANDLE(native_get_field_@(member.name)_message(messageHandle, i));
+          @[        end if]@
+        }
+      }
+
+
```
Comment (theseankelly) [inline rosidl_generator_dotnet/resource/msg.cs.em:228]: remove extra lines

File: rosidl_generator_dotnet/rosidl_generator_dotnet/__init__.py [hunk 5]
```diff
@@ -159,3 +164,22 @@ def get_field_name(type_name, field_name):
         return "{0}_".format(type_name)
     else:
         return upperfirst(field_name)
+
+
+BASIC_DOTNET_TYPES_TO_BASIC_IDL_TYPES = {
```
Comment (theseankelly) [inline rosidl_generator_dotnet/rosidl_generator_dotnet/__init__.py:169]: I can't find where this is used -- is it needed?

File: rosidl_generator_dotnet/rosidl_generator_dotnet/__init__.py [hunk 6]
```diff
@@ -133,6 +133,11 @@ def get_builtin_dotnet_type(type_, use_primitives=True):
 
     assert False, "unknown type '%s'" % type_
 
+def get_idl_type(type_):
```
Comment (theseankelly) [inline rosidl_generator_dotnet/rosidl_generator_dotnet/__init__.py:136]: I can't find where this is used -- is it needed?

File: rosidl_generator_dotnet/resource/msg.c.em [hunk 7]
```diff
@@ -1,5 +1,7 @@
 @{
 from rosidl_generator_dotnet import msg_type_to_c
+from rosidl_generator_dotnet import get_builtin_dotnet_type
```
Comment (theseankelly) [inline rosidl_generator_dotnet/resource/msg.c.em:3]: not used

File: rosidl_generator_dotnet/resource/msg.c.em [hunk 8]
```diff
@@ -1,5 +1,7 @@
...
+from rosidl_generator_dotnet import get_idl_type
```
Comment (theseankelly) [inline rosidl_generator_dotnet/resource/msg.c.em:4]: not used

**Comment by theseankelly** [comment]
Looks like the unit tests failed in the Windows build -- can you have a look and see if changes are needed @fmrico?

**Comment by Camelron** [comment]
Just so you know @fmrico , I have adapted the code in these commits to work with primitive sequences. We plan to pull the sequence support into the ms-iot fork of this repository and then eventually into this repository.

**Comment by esteve** [comment]
@Camelron @ooeygui any chance the changes in your fork will be upstreamed? They seem rather isolated and should be easy to review on our side. Thanks.

**Comment by ooeygui** [comment]
Hello @esteve ,
Absolutely! I sent you a note offline.
Our work in ms-iot is about high bandwidth staging of changes which we will upstream when in good states. We will need to resolve our changes with @fmrico as our changes build on theirs.

File: rosidl_generator_dotnet/resource/msg.c.em [hunk 9]
```diff
@@ -46,7 +48,49 @@ const void * @(msg_typename)__get_typesupport() {
...
+////////////////////////////////////////////////////////
+// DOING: Array types support
+
+void * @(msg_typename)__get_field_@(member.name)_message(void *message_handle, int index) {
+  @(msg_typename) * ros_message = (@(msg_typename) *)message_handle;
+  return &(ros_message->@(member.name)[index]);
+}
+
+int @(msg_typename)__getsize_array_field_@(member.name)_message()
+{
```
[the remark this answers is not in the document]
Comment (fmrico) [reply rosidl_generator_dotnet/resource/msg.c.em:60]: done

File: rosidl_generator_dotnet/resource/msg.c.em [hunk 10]
```diff
@@ -46,7 +48,49 @@ const void * @(msg_typename)__get_typesupport() {
 
 @[for member in message.structure.members]@
 @[    if isinstance(member.type, Array)]@
-// TODO: Array types are not supported
+////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (fmrico) [reply rosidl_generator_dotnet/resource/msg.c.em:51]: done

File: rosidl_generator_dotnet/resource/msg.cs.em [hunk 11]
```diff
@@ -34,7 +35,11 @@ public class @(type_name) : IMessage {
     {
 @[for member in message.structure.members]@
 @[    if isinstance(member.type, Array)]@
-// TODO: Array types are not supported
+@[        if isinstance(member.type.value_type, BasicType)]@       
+        // @(member.type.value_type) @(member.type.value_type.typename)
```
[the remark this answers is not in the document]
Comment (fmrico) [reply rosidl_generator_dotnet/resource/msg.cs.em:39]: I don't find it. Maybe I removed

File: rosidl_generator_dotnet/resource/msg.cs.em (hunk 1 above)
Comment (fmrico) [reply rosidl_generator_dotnet/resource/msg.cs.em:141]: done

File: rosidl_generator_dotnet/resource/msg.cs.em (hunk 3 above)
Comment (fmrico) [reply rosidl_generator_dotnet/resource/msg.cs.em:209]: Done

File: rosidl_generator_dotnet/resource/msg.cs.em (hunk 4 above)
Comment (fmrico) [reply rosidl_generator_dotnet/resource/msg.cs.em:228]: Done

File: rosidl_generator_dotnet/rosidl_generator_dotnet/__init__.py (hunk 5 above)
Comment (fmrico) [reply rosidl_generator_dotnet/rosidl_generator_dotnet/__init__.py:169]: Done

**Comment by fmrico** [comment]
Hi @esteve @theseankelly 

I could dedicate some cycles to continue with this. I have resolved the @theseankelly 's review comments. Now also works in Foxy.

Tests are passing for me, except for dotnet_cmake_module (lint_cmake), but this is external to this repo

@ooeygui I hope this PR will be useful for you. I hope to help as soon as this PR closes.

Best

**Comment by fmrico** [comment]
@theseankelly @esteve Could you check CI? It seems to be failing for something related to disabled commands in CI (add-path, setenv).

**Pull request merged**